lib/moment_matching_da.py

```diff
@@ -0,0 +1,153 @@
+"""Domain adaptation via moment matching.
+"""
+import tensorflow as tf
+import tensorflow.keras as tkf
+
+class MomentMatchingClassifier(tkf.Model):
+  """Moments matching classifier for unsupervised domain adaptation.
+
+  Arguments:
+  classifier: classifier model.
+  target_domain: target domain name.
+  num_layers: number of resblock layers in the classifier.
+  """
+  def __init__(self, classifier, domains, target_domain,
+               num_layers, **args):
+    super(MomentMatchingClassifier, self).__init__(**args)
+    self.classifier = classifier
+    self.domains = domains
+    self.num_domains = len(self.domains)
+    self.target_domain = target_domain
+    if target_domain in domains:
+      self.target_domain_idx = int(self.domains.index(self.target_domain))
+    else:
+      raise ValueError("The target domain is not in the domains list.")
+
+    self.num_layers = num_layers
+    self._input_shape = self.classifier.input_shape[0][1:]
+    self._set_feature_extractor()
+
+
+  def _set_feature_extractor(self):
```

Comment:
Looks like quite some code is present (duplicated) in both moment matching and discriminator da.